pipeline.cpp

```diff
@@ -81,6 +92,38 @@ void DrawLine(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* cons
 void DrawTriangle(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* const attrs, Attributes* const uniforms, FragmentShader* const frag)
 {
     // Your code goes here
+    // Code adapted from www.sunshine2k.de/coding/java/TriangleRasterization/TriangleRasterization.html
+    // This is known as the BaryCentric Algorithm. 
+    // Find the bounding box of the triangle, test each point in the box to see if it is in the triangle, if so, draw that pixel.
+    int maxX = fmax(triangle[0].x, fmax(triangle[1].x, triangle[2].x)); // Find the max x vertex by searching through the verticies of the triangle
+    int minX = fmin(triangle[0].x, fmin(triangle[1].x, triangle[2].x)); // Find the min x vertex by searching through the verticies of the triangle
+    int maxY = fmax(triangle[0].y, fmax(triangle[1].y, triangle[2].y)); // Find the max y vertex by searching through the verticies of the triangle
+    int minY = fmin(triangle[0].y, fmin(triangle[1].y, triangle[2].y)); // Find the min y vertex by searching through the verticies of the triangle
+
+    // Create the verticies that we want to draw with
+    Vertex vs1 = {triangle[1].x - triangle[0].x, triangle[1].y - triangle[0].y, 1, 1};
+    Vertex vs2 = {triangle[2].x - triangle[0].x, triangle[2].y - triangle[0].y, 1, 1};
+
+    // Start looping through the bounding box to determine what to draw
+    for (int y = minY; y <= maxY; ++y)
+    {
+        for (int x = minX; x <= maxX; ++x)
+        {
+            // Create vertex used for the Cross Products
+            Vertex q = {x - triangle[0].x, y - triangle[0].y, 1, 1};
+
+            // Find the Cross Products needed to test location of the pixel in question
+            float s = (float)crossProduct(q, vs2) / crossProduct(vs1, vs2);
```

Comment:
Cross Product function returns a float, so you don't need to cast the result to float data type